src/wallet/wallet.cpp

```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+            data->solvable_wallet = CreateWallet(context, wallet_name, std::nullopt, options, status, error, warnings);
+            if (status != DatabaseStatus::SUCCESS) {
+                error = _("Error: Failed to create new watchonly wallet");
+                return false;
+            }
+            created_wallets.push_back(data->solvable_wallet);
+            LOCK(data->solvable_wallet->cs_wallet);
+
+            // Parse the descriptors and add them to the new wallet
+            for (const auto& [desc_str, creation_time] : data->solvable_descs) {
+                // Parse the descriptor
+                FlatSigningProvider keys;
+                std::string parse_err;
+                std::unique_ptr<Descriptor> desc = Parse(desc_str, keys, parse_err, /* require_checksum */ true);
+                assert(desc); // It shouldn't be possible to have the LegacyScriptPubKeyMan make an invalid descriptor
+                assert(!desc->IsRange()); // It shouldn't be possible to have LegacyScriptPubKeyMan make a ranged watchonly descriptor
+
+                // Add to the wallet
+                WalletDescriptor w_desc(std::move(desc), creation_time, 0, 0, 0);
+                data->solvable_wallet->AddWalletDescriptor(w_desc, keys, "", false);
+            }
+
+            // Add the wallet to settings
+            UpdateWalletSetting(*context.chain, wallet_name, /*load_on_startup=*/true, warnings);
+        }
+    }
+
+    // Add the descriptors to wallet, remove LegacyScriptPubKeyMan, and cleanup txs and address book data
+    if (!wallet.ApplyMigrationData(*data, error)) {
+        return false;
+    }
+    return true;
+}
+
+bool MigrateLegacyToDescriptor(std::shared_ptr<CWallet>&& wallet, WalletContext& context, bilingual_str& error)
+{
+    LOCK(wallet->cs_wallet);
+
+    // Make a backup of the DB
+    fs::path this_wallet_dir = fs::absolute(fs::PathFromString(wallet->GetDatabase().Filename())).parent_path();
+    fs::path backup_filename = fs::PathFromString(strprintf("%s-%d.legacy.bak", wallet->GetName(), GetTime()));
+    fs::path backup_path = this_wallet_dir / backup_filename;
+    if (!wallet->BackupWallet(fs::PathToString(backup_path))) {
+        error = _("Error: Unable to make a backup of your wallet");
+        return false;
+    }
+
+    // First change to using SQLite
+
+    // Remove this wallet from the list of wallets while we mess with the database
+    // This is a workaround to prevent MaybeCompactWalletDB from accessing m_database while we are deleting and changing it
+    {
+        LOCK(context.wallets_mutex);
+        std::vector<std::shared_ptr<CWallet>>::iterator i = std::find_if(context.wallets.begin(), context.wallets.end(), [&wallet](std::shared_ptr<CWallet> w) {
+                return w == wallet;
+            });
+        assert(i != context.wallets.end());
+        context.wallets.erase(i);
+    }
+    // Now do the database stuff
+    if (!wallet->MigrateToSQLite(error)) return false;
```

Comment:
Since these are filesystem errors, I don't think restoring the backup would actually work. Furthermore, because `m_database` has been swapped out, and potentially doesn't have a backing database, I don't think it's safe to continue execution in that state as the `CWallet` would still exist, but nowhere to write data to.